src/main/scala/scalismo/mesh/boundingSpheres/SurfaceIntersectionIndex.scala

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright 2016 University of Basel, Graphics and Vision Research Group
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package scalismo.mesh.boundingSpheres
+
+import scalismo.geometry.{ Dim, Point, Vector, _3D }
+import scalismo.mesh.{ BarycentricCoordinates, TriangleId, TriangleMesh3D }
+
+trait SurfaceIntersectionIndex[D <: Dim] {
+
+  def hasIntersection(point: Point[D], direction: Vector[D]): Boolean
+
+  def getIntersectionPoints(point: Point[D], direction: Vector[D]): Seq[Point[D]]
+
+}
+
+trait TriangulatedSurfaceIntersectionIndex[D <: Dim] extends SurfaceIntersectionIndex[D] {
+
+  def getSurfaceIntersectionPoints(point: Point[D], direction: Vector[D]): Seq[(TriangleId, BarycentricCoordinates)]
+}
+
+object LineSurfaceIntersectionIndexBuilder {
+
+  /**
+   * Creates SurfaceDistance for a TriangleMesh3D.
+   */
+  def fromTriangleMesh3D(mesh: TriangleMesh3D): TriangulatedSurfaceIntersectionIndex[_3D] = {
+
+    // build triangle list (use only Vector[_3D], no Points)
+    val triangles = mesh.triangulation.triangles.map { t =>
+
+      val a = mesh.pointSet.point(t.ptId1).toVector
+      val b = mesh.pointSet.point(t.ptId2).toVector
+      val c = mesh.pointSet.point(t.ptId3).toVector
+      val ab = b - a
+      val ac = c - a
+
+      new Triangle(
+        a, b, c,
+        ab, ac,
+        ab.crossproduct(ac)
+      )
+
+    }
+
+    // build up search structure
+    val bs = BoundingSpheres.createForTriangles(triangles)
+
+    // new distance object
+    new LineTriangleMesh3DIntersectionIndex(bs, mesh, triangles)
+
+  }
+
+}
+
+private[mesh] case class LineTriangleMesh3DIntersectionIndex(private val boundingSphere: BoundingSphere,
```

Comment:
I suggest to remove `case` here: the class does not support the "value"-type equality and hash semantics one would expect due to the `BoundingSphere` member (usage in `TriangleMesh3DOperations` is already with `new` anyway)